### PR TITLE
WS and JPEG rework

### PR DESCRIPTION
**Pull request by Lu-Fi**
*fixed a bug in reading multiple osd values 
*fixed a bug in read and write motion roi
*fixed a bug receiving large json data
*fixed a bug in separator handling (,) in JSON response

Websocket rework:
The websocket implementation is now session capable. 
Incomming messages are now stored in the session specific structure 'u_ctx' in the member 'rx_message'.
Outgoing messages are now stored in the session specific structure 'u_ctx' in the member 'tx_message'.

When the preview image is requested via websocket, no current image is now requested from the encoder, it is always read and sent from /tmp/snapshot.jpg.
If the snapshot is requested faster than the encoder generates images, the delivery of the images is delayed on the server side.

Information on fps (frames per second) and Bps (bytes per second) can now be queried with json via websocket. 
{"stream(0|1|2)":{"stats":null}}
Inactive streams will return 0 values

The jpeg channel option 'jpeg_channel' which determines which encoder channel is used for the JPEG stream can now be configured with JSON via websocket.
{"stream2":{"jpeg_channel":null}}

The general options 'imp_polling_timeout' and 'osd_pool_size' also now available via json over websocket / http.
{"general":{"imp_polling_timeout":null, "osd_pool_size":null}}

WS now has set security required by default.
HTTP now has set security required by default.

'websocket.usertoken' option in prudynt.cfg now can be used to set a user specific token to acces websocket or http ressources.
This token must been added to the URL as a parameter.
new WebSocket('ws://<camera>:<port>?token=<usertoken>')
http://<camera>:<port>/preview.jpg?token=<usertoken>
sample: http://myThinginoCam:8089/preview.jpg?token=bigSecret

JPEG rework:
the JPEG thread continuously writes a current image to “/tmp/snapshot.jpg” when activated. 
If the image is requested via websocket or http, this file is always read and sent.

In preparation for an additional JPEG channel, the global variable 'gloabl_jpeg' is now an array.

If no preview images are requested via websocket or http, the jpeg thread goes into standby.
The channel is reactivated when new requests are received. 
The associated video stream is then also activated if it is not active.
Since the first encoder images after the start may be too dark, incomplete or without OSD, 
the first image is delivered with a delay after the thread is woken up. By default this is 100 milliseconds. 

Video rework:
The power saving option has been removed as the JPEG channel now activates the required video channel when needed. 
This means that the video channel can now always be suspended when no client is connected.

Config rework:
JPEG stream(2) 'format' option is removed from .cfg.
JPEG stream(2) 'height', 'width' option is removed from .cfg.

### CODE REVIEW

File: src/main.cpp [hunk 1]
```diff
@@ -93,7 +95,9 @@ int main(int argc, const char *argv[])
 
     global_video[0] = std::make_shared<video_stream>(0, &cfg->stream0, "stream0");
     global_video[1] = std::make_shared<video_stream>(1, &cfg->stream1, "stream1");
-    global_jpeg = std::make_shared<jpeg_stream>(jpeg_stream{2, &cfg->stream2});
+    global_jpeg[0] = std::make_shared<jpeg_stream>(2, &cfg->stream2);
+    //global_jpeg[1] = std::make_shared<jpeg_stream>(jpeg_stream{3, &cfg->stream3});
```
Comment (nschimme) [inline src/main.cpp:99]: Is this line necessary?

File: src/worker.cpp [hunk 2]
```diff
@@ -143,37 +75,72 @@ void *Worker::jpeg_grabber(void *arg)
     LOG_DEBUG("Start jpeg_grabber thread.");
 
     StartHelper *sh = static_cast<StartHelper *>(arg);
-    int encChn = sh->encChn;
+    int jpgChn = sh->encChn - 2;
```
Comment (nschimme) [inline src/worker.cpp:78]: Why -2? Can we calculate this from the stream number?

File: src/worker.cpp [hunk 3]
```diff
@@ -143,37 +75,72 @@ void *Worker::jpeg_grabber(void *arg)
...
+
+    global_jpeg[jpgChn]->imp_encoder = IMPEncoder::createNew(
+        global_jpeg[jpgChn]->stream, sh->encChn, global_jpeg[jpgChn]->stream->jpeg_channel, "stream2");
 
     // inform main that initialization is complete
     sh->has_started.release();
 
-    ret = IMP_Encoder_StartRecvPic(global_jpeg->encChn);
-    LOG_DEBUG_OR_ERROR(ret, "IMP_Encoder_StartRecvPic(" << global_jpeg->encChn << ")");
+    ret = IMP_Encoder_StartRecvPic(global_jpeg[jpgChn]->encChn);
```
Comment (nschimme) [inline src/worker.cpp:95]: We have `0` hard coded sometimes but here we use `jpgChn`. Should we fix the inconsistencies?

File: src/worker.cpp [hunk 4]
```diff
@@ -394,13 +419,28 @@ void *Worker::stream_grabber(void *arg)
                 LOG_DDEBUG("IMP_Encoder_PollingStream(" << encChn << ", " << cfg->general.imp_polling_timeout << ") timeout !");
             }
         }
-        else
+        else if(global_video[encChn]->onDataCallback == nullptr && !global_restart_video && !global_video[encChn]->run_for_jpeg)
```
Comment (nschimme) [inline src/worker.cpp:422]: missing space between `if(`

File: src/WS.cpp [hunk 5]
```diff
@@ -232,14 +263,20 @@ enum
...
+    PNT_STREAM2_STATS,
+    PNT_STREAM2_FPS
 };
 
 static const char *const stream2_keys[] = {
     "jpeg_enabled",
     "jpeg_path",
     "jpeg_quality",
-    "jpeg_refresh"};
+    //"jpeg_refresh",
```
Comment (nschimme) [inline src/WS.cpp:276]: Remove this line instead?

File: src/WS.cpp [hunk 6]
```diff
@@ -207,12 +234,15 @@ enum
     PNT_STREAM_ROTATION,
     PNT_STREAM_SCALE_WIDTH,
     PNT_STREAM_SCALE_HEIGHT,
+    PNT_STREAM_STATS,
     PNT_STREAM_OSD
 };
 
 static const char *const stream_keys[] = {
```
Comment (nschimme) [inline src/WS.cpp:241]: Are there any equivalent keys we need to update for web sockets? i.e. websocket.usertoken

File: src/WS.cpp [hunk 7]
```diff
@@ -1112,27 +1209,40 @@ signed char WS::stream_callback(struct lejp_ctx *ctx, char reason)
...
+                cfg->set<bool>(u_ctx->path, false);
+            }
+            append_session_msg(
+                u_ctx->message, "%s", cfg->get<bool>(u_ctx->path) ? "true" : "false");
         }
         else
         {
             switch (ctx->path_match)
-            {
+            {
```
Comment (nschimme) [inline src/WS.cpp:1240]: Extra trailing whitespace that needs to trimmed

File: src/WS.cpp [hunk 8]
```diff
@@ -1217,26 +1351,63 @@ signed char WS::stream2_callback(struct lejp_ctx *ctx, char reason)
                 {
                 }
             }
-            append_message(
-                "%d", cfg->get<int>(u_ctx->path));
+            append_session_msg(
+                u_ctx->message, "%d", cfg->get<int>(u_ctx->path));
             break;
+        /*
         case PNT_STREAM2_JPEG_REFRESH:
```
Comment (nschimme) [inline src/WS.cpp:1358]: Do we need this commented out code path here still? It's probably out of date now.

File: src/WS.cpp [hunk 9]
```diff
@@ -1217,26 +1351,63 @@ signed char WS::stream2_callback(struct lejp_ctx *ctx, char reason)
...
+            }
+            append_session_msg(
+                u_ctx->message, "%d", cfg->get<int>(u_ctx->path));
             break;
+        case PNT_STREAM2_STATS:
+            if (reason == LEJPCB_VAL_NULL)
+            {
+                uint8_t fps = 0;
+                uint32_t bps = 0;
+                if (u_ctx->root == "stream2")
```
Comment (nschimme) [inline src/WS.cpp:1394]: Why do we hardcode `stream2` here?

File: src/main.cpp (hunk 1 above)
Comment (Lu-Fi) [reply src/main.cpp:99]: in preparation to have more than 1 JPEG channel in future this placeholder was added and should not harm.
And no it's not necessary.

File: src/worker.cpp (hunk 2 above)
Comment (Lu-Fi) [reply src/worker.cpp:78]: encChn is the stream number and -2 is a calculation ?

File: src/WS.cpp [hunk 10]
```diff
@@ -1773,245 +1983,438 @@ signed char WS::root_callback(struct lejp_ctx *ctx, char reason)
...
+        if (flag & PNT_FLAG_RESTART_AUDIO)
+        {
+            global_restart_audio = true;
+            flag &= ~PNT_FLAG_RESTART_AUDIO;
+        }
+        global_cv_worker_restart.notify_one();
+    }
+}
+
+std::string generateSessionID()
```
Comment (nschimme) [inline src/WS.cpp:2012]: So far we've usually been using `Id` for other identifiers.

File: src/WS.cpp [hunk 11]
```diff
@@ -1773,245 +1983,438 @@ signed char WS::root_callback(struct lejp_ctx *ctx, char reason)
...
-    int ws_send_msg_length;
-    char *url_ptr;
+
+    // security token ?token=
     char url_token[128]{0};
     char content_type[128]{0};
     std::string json_data((char *)in, len);
 
-    if (reason == LWS_CALLBACK_ESTABLISHED || reason == LWS_CALLBACK_HTTP)
+    //LOG_DDEBUG(reason);
```
Comment (nschimme) [inline src/WS.cpp:2049]: Remove this commented out code?

File: src/worker.cpp (hunk 3 above)
Comment (Lu-Fi) [reply src/worker.cpp:95]: in preparation to have more than 1 JPEG channel in future.

File: src/WS.cpp [hunk 12]
```diff
@@ -1773,245 +1983,438 @@ signed char WS::root_callback(struct lejp_ctx *ctx, char reason)
...
+            " ,flag:" << u_ctx->flag << 
+            " ,ip:" << client_ip << 
+            " ,len:" << len << 
+            " ,last:" << lws_is_final_fragment(wsi));
+
+        /* larger requests can be segmented into several requests, 
+         * so we have to collect all the data until we reach the last segment.
+         * On receiving the first segment we should clear the rx_message
+         */
+        if(lws_is_first_fragment(wsi))
```
Comment (nschimme) [inline src/WS.cpp:2096]: Missing space between `if(`

File: src/WS.cpp [hunk 13]
```diff
@@ -1773,245 +1983,438 @@ signed char WS::root_callback(struct lejp_ctx *ctx, char reason)
...
+            u_ctx->rx_message.clear();
+
+        u_ctx->rx_message.append(json_data);
+
+        if(!lws_is_final_fragment(wsi))
```
Comment (nschimme) [inline src/WS.cpp:2101]: missing space here too

File: src/WS.cpp [hunk 14]
```diff
@@ -1773,245 +1983,438 @@ signed char WS::root_callback(struct lejp_ctx *ctx, char reason)
...
+                lws_write(wsi, jpeg_buf.data() + LWS_PRE, jpeg_buf.size() - LWS_PRE, LWS_WRITE_BINARY);
+            }
+            u_ctx->flag &= ~(PNT_FLAG_WS_SEND_PREVIEW | PNT_FLAG_WS_PREVIEW_PENDING);
         }
 
-        // always reset signal
-        u_ctx->signal = 0;
-        
+        // cleanup request pending flag
+        //u_ctx->flag &= ~PNT_FLAG_WS_REQUEST_PENDING;
```
Comment (nschimme) [inline src/WS.cpp:2222]: Still needed?

File: src/worker.cpp (hunk 2 above)
Comment (nschimme) [reply src/worker.cpp:78]: Yes, I meant as in is it always `stream2` What if its `stream3 in the future?

File: src/WS.cpp (hunk 5 above)
Comment (Lu-Fi) [reply src/WS.cpp:276]: will be done in future, if i'am really sure we do not need jpeg_refresh.
This value now is replaced by fps.

File: src/worker.cpp (hunk 4 above)
Comment (Lu-Fi) [reply src/worker.cpp:422]: pull incomming

File: src/WS.cpp (hunk 6 above)
Comment (Lu-Fi) [reply src/WS.cpp:241]: For security reasons I have not implemented websocket configuration customizations via websocket or html.

File: src/WS.cpp (hunk 7 above)
Comment (Lu-Fi) [reply src/WS.cpp:1240]: pull incomming.

File: src/WS.cpp (hunk 8 above)
Comment (Lu-Fi) [reply src/WS.cpp:1358]: will be done in future, if i'am really sure we do not need jpeg_refresh.
This value now is replaced by fps.

File: src/WS.cpp (hunk 9 above)
Comment (Lu-Fi) [reply src/WS.cpp:1394]: in preparation to have more than 1 JPEG channel in future.

File: src/WS.cpp (hunk 10 above)
Comment (Lu-Fi) [reply src/WS.cpp:2012]: first use of 'id' in this project, but can be renamed to session_id. Whats your opinion on this ? @gtxaspec , @themactep

File: src/WS.cpp (hunk 11 above)
Comment (Lu-Fi) [reply src/WS.cpp:2049]: pull incomming

File: src/WS.cpp (hunk 10 above)
Comment (nschimme) [reply src/WS.cpp:2012]: Yeah, I meant in comparison with `devId` or `clientSessionId` that already exist in the code from a code style naming perspective.

Comment (nschimme) [reply src/WS.cpp:2012]: In relation to the `clientSessionId`, maybe we should make it more clear this is a `wsSessionId` ?

File: src/WS.cpp (hunk 14 above)
Comment (Lu-Fi) [reply src/WS.cpp:2222]: pull incomming

File: src/WS.cpp (hunk 13 above)
Comment (Lu-Fi) [reply src/WS.cpp:2101]: pull incomming

File: src/WS.cpp (hunk 12 above)
Comment (Lu-Fi) [reply src/WS.cpp:2096]: pull incomming

File: src/worker.cpp (hunk 2 above)
Comment (Lu-Fi) [reply src/worker.cpp:78]: We currently have encChn as an encoder channel.
This is also used as an index for the global_video array. 
A JPEG channel is also an encoder channel but due to the structure in a different array (global_jpeg) if we don't want to start with the global_jpeg array at 2 I have to subtract the video channels. 
But I am open to suggestions.

**Pull request merged**